### PR TITLE
Automatically test all DTOs

### PR DESCRIPTION
**Pull request by qparis**

### CODE REVIEW

**Comment by plata** [comment]
The PR has many changes which are not related to this PR. This makes it quite difficult to review it properly. We do not have this issue in other reviews so I don't quite understand why this happens for your last PRs.

**Comment by qparis** [comment]
Don't understand either... 
I think I need to refork the whole project. The only files I have changed are those whose names contain DTO

**Comment by plata** [comment]
So this fixes #686 as well? (Is it just me or is the GitHub search broken?)

**Comment by qparis** [comment]
Yep it does fix this :)

**Comment by plata** [comment]
Ok. Then I would like @madoar to do the review because he opened that issue.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/library/LibraryPanel.java [hunk 1]
```diff
@@ -118,6 +122,10 @@ public void setShortcutDTO(ShortcutDTO shortcutDTO) {
         this.setCenter(vBox);
     }
 
+    private String unCamelize(String s) {
```
Comment (madoar) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/library/LibraryPanel.java]: What exactly does this method? 
Some JavaDoc would be nice here

File: phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+import java.net.URISyntaxException;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.junit.Assert.*;
+
+@RunWith(Parameterized.class)
+public class DTOContractTest {
+    private final static int DEFAULT_MAX_DEPTH = 5;
```
Comment (madoar) [inline phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java]: For what is this variable used?
I guess for some kind of recursion?
Again some JavaDoc would be nice.

File: phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+        return null;
+    }
+
+    /**
+     * This aim of the class is to customize the content of
+     * the name of the test suit
+     *
+     * @param <C> The class the
+     */
```
Comment (madoar) [inline phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java]: Please check the JavaDoc of this class again.

File: phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        assertTrue(json.endsWith("}"));
+    }
+
+    @Test
+    public void testDeserialize() throws ReflectiveOperationException, IOException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        final String json = objectMapper.writeValueAsString(dto);
+
+        final ObjectMapper alternativeObjectMapper = new ObjectMapper();
+        alternativeObjectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```
Comment (madoar) [inline phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java]: This sounds like we're allowing unknown properties in the json, when we want to deserialize it.
Do we also allow this in production?

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/library/LibraryPanel.java (hunk 1 above)
Comment (qparis) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/library/LibraryPanel.java]: This is not related to this PR. 
See @plata comment

**Comment by qparis** [comment]
Sure!
Can we merge this first to allow others PR to be merged? (My fork is messed up)
I will add extra comments in a separate pull request

File: phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        final Object dto1 = newDTOInstance(DTOContainer.getClazz());
+        final Object dto2 = newDTOInstance(DTOContainer.getClazz());
+        assertEquals(dto1, dto2);
+        assertEquals(dto1.hashCode(), dto2.hashCode());
+    }
+
+    @Test
+    public void testEqualsTrivialCases() throws ReflectiveOperationException, URISyntaxException {
+        final Object dto = newDTOInstance(DTOContainer.getClazz());
+        assertTrue(dto.equals(dto));
```
Comment (madoar) [inline phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java]: Just asking:
Is there a difference between: `assertTrue(dto.equals(dto));` and `assertEquals(dto,  dto);`?

Comment (qparis) [inline phoenicis-tests/src/test/java/org/phoenicis/tests/DTOContractTest.java]: No fondamental differences but in this particular case, we are testing the equals method. So we really want to call it explicitely
